### PR TITLE
Backport of INT changes 

### PR DESCRIPTION
**Pull request by MatthewSandfordImprobable**
4.26 Engine PR: https://github.com/improbableio/UnrealEngine/pull/1068
4.27 Engine PR: https://github.com/improbableio/UnrealEngine/pull/1064

### CODE REVIEW

File: SpatialGDK/Source/SpatialGDK/Public/Utils/SpatialActorUtils.h [hunk 1]
```diff
@@ -85,13 +86,9 @@ inline FVector GetActorSpatialPosition(const AActor* InActor)
 	}
 	else if (const APlayerController* PlayerController = Cast<APlayerController>(Controller))
 	{
-		if (PlayerController->IsInState(NAME_Spectating))
+		if (USceneComponent* RootComponent = PlayerController->GetRootComponent())
```
Comment (MatthewSandfordImprobable) [inline SpatialGDK/Source/SpatialGDK/Public/Utils/SpatialActorUtils.h:89]: This change might be something to talk about?

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp [hunk 2]
```diff
@@ -688,7 +688,7 @@ void USpatialNetDriver::CreateAndInitializeLoadBalancingClasses()
 	const TSubclassOf<UAbstractSpatialMultiWorkerSettings> MultiWorkerSettingsClass =
 		USpatialStatics::GetSpatialMultiWorkerClass(CurrentWorld);
 
-	const UAbstractSpatialMultiWorkerSettings* MultiWorkerSettings =
+	UAbstractSpatialMultiWorkerSettings* MultiWorkerSettings =
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp:691]: Doesn't look like we need this change, lets revert.

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp [hunk 3]
```diff
@@ -2853,6 +2860,7 @@ void USpatialNetDriver::PostSpawnPlayerController(APlayerController* PlayerContr
 	PlayerController->SetReplicates(true);
 	PlayerController->Role = OriginalRole;
 	PlayerController->SetPlayer(OwnershipConnection);
+	PlayerController->ChangeState(NAME_Playing);
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp:2863]: @sjtsilvester context on this? This will get called when the pawn possesses a pawn, so not sure why we need to manually call it here.

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp [hunk 4]
```diff
@@ -3297,7 +3315,7 @@ void USpatialNetDriver::TryFinishStartup()
 			const TSubclassOf<UAbstractSpatialMultiWorkerSettings> MultiWorkerSettingsClass =
 				USpatialStatics::GetSpatialMultiWorkerClass(GetWorld());
 
-			const UAbstractSpatialMultiWorkerSettings* MultiWorkerSettings =
+			UAbstractSpatialMultiWorkerSettings* MultiWorkerSettings =
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp:3318]: likewise here

File: SpatialGDK/Source/SpatialGDK/Private/Interop/ClientConnectionManager.cpp [hunk 5]
```diff
@@ -76,6 +77,30 @@ void ClientConnectionManager::OnRequestReceived(const Worker_Op&, const Worker_C
 void ClientConnectionManager::RegisterClientConnection(const Worker_EntityId InWorkerEntityId, USpatialNetConnection* ClientConnection)
 {
 	WorkerConnections.Add(InWorkerEntityId, ClientConnection);
+
+	const EntityViewElement* EntityView = SubView->GetView().Find(InWorkerEntityId);
+	if (!ensureAlwaysMsgf(EntityView != nullptr,
+						  TEXT("Failed to entity component data for system worker entity %lld. Client IP will be unset."),
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/ClientConnectionManager.cpp:83]:
```suggestion
						  TEXT("Failed to find entity component data for system worker entity %lld. Client IP will be unset."),
```

File: SpatialGDK/Source/SpatialGDK/Private/Interop/ClientConnectionManager.cpp [hunk 6]
```diff
@@ -76,6 +77,30 @@ void ClientConnectionManager::OnRequestReceived(const Worker_Op&, const Worker_C
 void ClientConnectionManager::RegisterClientConnection(const Worker_EntityId InWorkerEntityId, USpatialNetConnection* ClientConnection)
 {
 	WorkerConnections.Add(InWorkerEntityId, ClientConnection);
+
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/ClientConnectionManager.cpp:80]: This section is in this PR also - https://github.com/spatialos/UnrealGDK/pull/3286/files.
Reminder to close that other PR when this goes in.

File: SpatialGDK/Source/SpatialGDK/Private/Interop/ClientConnectionManager.cpp [hunk 7]
```diff
@@ -76,6 +77,30 @@ void ClientConnectionManager::OnRequestReceived(const Worker_Op&, const Worker_C
...
+						  TEXT("Failed to access system worker component data for system worker entity %lld. Client IP will be unset."),
+						  InWorkerEntityId))
+	{
+		return;
+	}
+
+	const SpatialGDK::Worker WorkerData(Data->GetUnderlying());
+	ClientConnection->ClientIP = *WorkerData.Connection.IPAddress;
+
+	UE_LOG(LogTemp, Log, TEXT("Registered client connection. System entity: %lld. Client IP: %s."), InWorkerEntityId,
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/ClientConnectionManager.cpp:102]:
```suggestion
	UE_LOG(LogWorkerEntitySystem, Log, TEXT("Registered client connection. System entity: %lld. Client IP: %s."), InWorkerEntityId,
```

File: SpatialGDK/Source/SpatialGDK/Private/Interop/Connection/SpatialConnectionManager.cpp [hunk 8]
```diff
@@ -138,6 +138,7 @@ struct ConfigureConnection
 		UnrealGDKVersionPair.name = "gdk_version";
 		UnrealGDKVersionPair.version = GDKVersion.Get();
 		Params.versions = &UnrealGDKVersionPair;
+		Params.network.connection_timeout_millis = 600000;
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/Connection/SpatialConnectionManager.cpp:141]: driveby - can we make this a spatial gdk config setting?

File: SpatialGDK/Source/SpatialGDK/Private/LoadBalancing/LayeredLBStrategy.cpp [hunk 9]
```diff
@@ -149,8 +158,11 @@ VirtualWorkerId ULayeredLBStrategy::WhoShouldHaveAuthority(const AActor& Actor)
 
 	const VirtualWorkerId ReturnedWorkerId = LayerNameToLBStrategy[LayerName]->WhoShouldHaveAuthority(*RootOwner);
 
-	UE_LOG(LogLayeredLBStrategy, Log, TEXT("LayeredLBStrategy returning virtual worker id %d for Actor %s."), ReturnedWorkerId,
-		   *AActor::GetDebugName(RootOwner));
+	if (ReturnedWorkerId != SpatialConstants::INVALID_VIRTUAL_WORKER_ID)
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Private/LoadBalancing/LayeredLBStrategy.cpp:161]: @sjtsilvester context here?

File: SpatialGDK/Source/SpatialGDK/Public/Utils/SpatialActorUtils.h (hunk 1 above)
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Public/Utils/SpatialActorUtils.h:89]: @sjtsilvester this seems like a pretty big change, context would be good

File: SpatialGDK/Source/SpatialGDK/Public/Utils/SpatialActorUtils.h [hunk 10]
```diff
@@ -122,9 +119,35 @@ inline bool DoesActorClassIgnoreVisibilityCheck(AActor* InActor)
...
+		return false;
+	}
+
+	if (USpatialNetDriver* SpatialNetDriver = Cast<USpatialNetDriver>(World->GetNetDriver()))
+	{
+		if (SpatialNetDriver->GetReplicationDriver() != nullptr)
+		{
+			return true;
+		}
+		UE_LOG(LogTemp, Log, TEXT("Actor %s UsingSpatialRepGraph SpatialNetDriver->GetReplicationDriver() != nullptr"),
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Public/Utils/SpatialActorUtils.h:137]:
```suggestion
		UE_LOG(LogTemp, Log, TEXT("Actor %s UsingSpatialRepGraph SpatialNetDriver->GetReplicationDriver() == nullptr"),
```

File: SpatialGDK/Source/SpatialGDK/Private/Interop/ClientConnectionManager.cpp (hunk 6 above)
Comment (MatthewSandfordImprobable) [reply SpatialGDK/Source/SpatialGDK/Private/Interop/ClientConnectionManager.cpp:80]: Well pointed out. Will do!

File: SpatialGDK/Source/SpatialGDK/Public/Utils/SpatialActorUtils.h (hunk 1 above)
Comment (sjtsilvester) [reply SpatialGDK/Source/SpatialGDK/Public/Utils/SpatialActorUtils.h:89]: This was a change Intrepid made that we didn't catch at the time. The CL for this has the comment 

"Fixing snap to ground on server and remotes.
Tiny fix for SiegeTeleport to forcenetupdate when assigned siege team type"

but still unsure - they had an issue where players would teleport back to their original location when being teleported to the siege island so it may have been an attempt to fix that. I will ask.

File: SpatialGDK/Source/SpatialGDK/Private/LoadBalancing/LayeredLBStrategy.cpp [hunk 11]
```diff
@@ -45,14 +45,23 @@ FString ULayeredLBStrategy::ToString() const
...
-void ULayeredLBStrategy::SetLayers(const TArray<FLayerInfo>& WorkerLayers)
+void ULayeredLBStrategy::SetLayers(TArray<FLayerInfo>& WorkerLayers)
 {
 	check(WorkerLayers.Num() != 0);
 
 	// For each Layer, add a LB Strategy for that layer.
 	for (int32 LayerIndex = 0; LayerIndex < WorkerLayers.Num(); ++LayerIndex)
 	{
-		const FLayerInfo& LayerInfo = WorkerLayers[LayerIndex];
+		FLayerInfo& LayerInfo = WorkerLayers[LayerIndex];
```
Comment (sjtsilvester) [inline SpatialGDK/Source/SpatialGDK/Private/LoadBalancing/LayeredLBStrategy.cpp:55]: Context for this - if the LBS for a layer becomes corrupted somehow (which has happened within Intrepid, possibly an artifact of P4 or engine upgrades), then without this change to set the LBS to the default, it crashes while launching the map. This is also the reason why the consts have been removed in the startup flow for the multi layer settings.

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp (hunk 3 above)
Comment (MatthewSandfordImprobable) [reply SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp:2863]: Do you think we should ask Intrepid about it?

File: SpatialGDK/Source/SpatialGDK/Public/Utils/SpatialActorUtils.h (hunk 1 above)
Comment (MatthewSandfordImprobable) [reply SpatialGDK/Source/SpatialGDK/Public/Utils/SpatialActorUtils.h:89]: Maybe this is a revert then?

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp (hunk 3 above)
Comment (sjtsilvester) [reply SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp:2863]: @m-samiec : Looks like we added this line to Intrepid at around the time of 0.13-RC to fix a bug. If the code that calls this when the pawn possesses a pawn was added after 0.13-RC, then I think we can remove this, otherwise I shall need to dig for the exact bug this was causing.

File: SpatialGDK/Source/SpatialGDK/Public/Utils/SpatialActorUtils.h (hunk 1 above)
Comment (sjtsilvester) [reply SpatialGDK/Source/SpatialGDK/Public/Utils/SpatialActorUtils.h:89]: Have asked Intrepid

Comment (sjtsilvester) [reply SpatialGDK/Source/SpatialGDK/Public/Utils/SpatialActorUtils.h]: Yep, this is a revert

File: CHANGELOG.md [hunk 12]
```diff
@@ -24,6 +24,9 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 - Users can now change the `Server Port` under `Editor Settings -> Level Editor -> Play -> Multiplayer Options`, or alternatively under `Advanced Settings` in the `Play` dropdown menu and launch local deployments that use the new `Server Port` value
 - Add Only Authoritative Server replication type. Variables marked with this type will only replicate to subsequent servers that gain authority of this actor. Can be accessed on standard `Replicated` variables by adding `COND_OnlyAuthServer` replication condition in `GetLifetimeReplicatedProps`.
 - Added a warning for using a `NetDeltaSerialized` property with PushModel due to Spatial GDK differing from Native Unreal in how `NetDeltaSerialized` properties are replicated.
+- Add a flag to allow RPCs to not be queued even with unresolved parameters, similar to the existing AllowUnresolvedParameters flag.
```
Comment (m-samiec) [inline CHANGELOG.md:27]:
```suggestion
- Add a function flag (SPATIALFUNC_NeverQueueForUnresolvedParameters) to allow RPCs to never queue even with unresolved parameters.
```

**Review by m-samiec — APPROVED** [review]

**Review by sjtsilvester — APPROVED** [review]
